recipes/qcoro/all/conanfile.py

```diff
@@ -23,124 +32,149 @@ class QCoroConan(ConanFile):
-            raise ConanInvalidConfiguration("imagl requires some C++20 features, which are available in libc++ for clang compiler.")
+            raise ConanInvalidConfiguration(
+                "qcoro requires some C++20 features, which are only available in libc++ for clang compiler."
+            )
 
         compiler_version = str(self.settings.compiler.version)
-
         minimum_version = self._compilers_minimum_version.get(str(self.settings.compiler), False)
-        if not minimum_version:
-            self.output.warn("qcoro requires C++20. Your compiler is unknown. Assuming it supports C++20.")
-        elif lazy_lt_semver(compiler_version, minimum_version):
-            raise ConanInvalidConfiguration("qcoro requires some C++20 features, which your {} {} compiler does not support.".format(str(self.settings.compiler), compiler_version))
-        else:
-            print("Your compiler is {} {} and is compatible.".format(str(self.settings.compiler), compiler_version))
+        if minimum_version and Version(compiler_version) < minimum_version:
+            raise ConanInvalidConfiguration(
+                f"qcoro requires some C++20 features, which your {str(self.settings.compiler)} "
+                f"{compiler_version} compiler does not support."
+            )
+
+    def build_requirements(self):
+        self.tool_requires("cmake/[>=3.23 <4]")
+        self.tool_requires("qt/<host_version>")
 
     def source(self):
-        tools.get(**self.conan_data["sources"][self.version],
-                  destination=self._source_subfolder, strip_root=True)
+        get(self, **self.conan_data["sources"][self.version], strip_root=True)
 
-    def _configure_cmake(self):
-        if self._cmake:
-            return self._cmake
+    @property
+    def _with_qml(self):
+        return self.dependencies["qt"].options.get_safe("qtdeclarative", False)
 
-        self._cmake = CMake(self)
+    @property
+    def _with_dbus(self):
+        return self.dependencies["qt"].options.get_safe("with_dbus", False)
 
-        self._cmake.definitions["QCORO_BUILD_EXAMPLES"] = False
-        self._cmake.definitions["QCORO_ENABLE_ASAN"] = self.options.asan
-        self._cmake.definitions["BUILD_TESTING"] = False
-        self._cmake.definitions["QCORO_WITH_QTDBUS"] = self.options["qt"].with_dbus
-        self._cmake.configure(build_folder=self._build_subfolder)
-        return self._cmake
+    @property
+    def _with_quick(self):
+        return (self.dependencies["qt"].options.get_safe("gui", False) and
+                self.dependencies["qt"].options.get_safe("qtshadertools", False))
+
+    @property
+    def _with_websockets(self):
+        return self.dependencies["qt"].options.get_safe("qtwebsockets", False)
+
+    def generate(self):
+        VirtualBuildEnv(self).generate()
+        # Required for Qt's moc and qtpaths
+        VirtualRunEnv(self).generate(scope="build")
```

Comment:
Do you really need to inject RunEnv in build scope ? Qt is also in build requirements, so I would have expected that this trick wouldn't be needed.

Reply:
After experimenting with adding cross-compilation support for the Qt6 recipe recently, I have a strong suspicion the exported Qt CMake modules in combination with the package_info() are not using the correct version of the tools and always rely on the `Qt6::moc` etc targets from the host profile instead. This explains why the VirtualRunEnv is necessary when Qt6 has been built with `shared=True.

I suppose the correct solution in this recipe for the time being would be to drop the tool_requires dependency of Qt instead.